Added rules `requireNamedUnassignedFunctions` and `disallowNamedUnassignedFunctions`. Fixes #655, #834.

twolfson opened the pull request:
As discussed in #655 and #834, there is no middle ground for requiring names on anonymous functions/not. These new rules aim to fix that by only requiring naming if the function has not been named by other means (i.e. no left hand assignment). In this PR:

- Added `requireNamedUnassignedFunctions` and `disallowNamedUnassignedFunctions` rules
- Added tests

**Missing:**

## Documentation
I couldn't find a place to put descriptions for the new rules. However, they would read something like:

### requireNamedUnassignedFunctions
Require inline name for any function that has not been assigned via a left-hand operator.

Type: `Boolean` or `Object`

Values:

- `true`
- `Object`:
    - `allExcept`: array of function names/methods to skip over if `function` is used as a parameter

**Example**

```js
"requireNamedUnassignedFunctions": { allExcept: ["describe", "describe.skip", "it", "it.skip"] }
```

**Valid**

```js
var foo = function() {};
[1].forEach(function iterateItems() {});
it(function() {}); // Valid because of `allExcept`
```

**Invalid**

```js
[1].forEach(function() {});
before(function() {});
```

### disallowNamedUnassignedFunctions
Disallow inline name for any function that has not been assigned via a left-hand operator.

Type: `Boolean`

Values: `true`

**Example**

```js
"disallowNamedUnassignedFunctions": true
```

**Valid**

```js
var foo = function() {};
[1].forEach(function() {});
before(function() {});
```

**Invalid**

```js
[1].forEach(function iterateItems() {});
before(function setupServer() {});
```

## Wildcards
It is possible people will want to use wildcards (e.g. `it.*`). However, this is an edge case on top of an edge case. It is hard to know exactly the use case so I have decided to defer implementation until someone brings it up as a necessity.

twolfson commented:
Ah sweet, noticing the `1.12` milestone. I am going to update this PR to fix those merge conflicts. I will comment when it's updated.

twolfson commented:
I have pushed an updated copy. The merge conflict was out of the addition for `pathval` as a dependency.

mikesherov commented:
@twolfson, thanks again for contributing! Looks like you have a slight regression in code coverage here. Please check the code coverage report by running: `npm run-script coverage-html && open coverage.html` to make sure you have tests for all cases. Thanks!

twolfson commented:
Yea, I noticed that. I am busy at the moment but will take a look by the end of Sunday.

twolfson commented:
I have fixed up the coverage issues and pushed the latest version.

zxqfox commented on lib/rules/require-named-unassigned-functions.js:
```diff
@@ -0,0 +1,84 @@
+        file.iterateNodesByType(['FunctionExpression'], function(node) {
+            var parentNode = node.parentNode;
+            if (!parentNode.type.match(/VariableDeclarator|Property|AssignmentExpression/)) {
+                if (node.id === null) {
```
`VariableDeclarator` nodes uses `id` property, `Property` uses `key`, `AssignmentExpression` uses `left`.

Absolutely not clear why we use here only `id` to check. ;-(

Can we use this structure?
```
// skip because...
if (parentNode.type.match...) {
  return;
}

// skip because
if (node.id === null) {
  return;
}

// ... the rest part of code
```

twolfson replied:
Sure thing. I was trying to keep comments light inline with the rest of the code but I will switch it up to your recommendations.

twolfson commented:
I have adjusted the code structure and added more comments as requested.

zxqfox commented:
@twolfson Wow. Thanks!

mikesherov commented:
@twolfson I know it's been a bit on this PR. Can you please rebase one last time here?

twolfson commented:
Based on my work in #1059, it might take me a bit of time to sort out the merge conflicts. I will rebase/sort it out by the end of the next weekend.

twolfson commented:
@mikesherov I have rebased to the latest `master`. Notable changes are:

- Relocated test files to `test/specs` for consistency
- Added documentation comments to rules for consistency
- Updated `configure` parameter to be named `options` for consistency

TheSavior commented:
This was tagged with 1.12, were we planning on landing this?

mikesherov commented:
Yes, we're going to land this soon. We need to start planning 1.13, 2.0, and make determinations on what the goals are. Can't land this until a potential 1.12.1 goes out also.

markelog commented:
> We need to start planning 1.13, 2.0, and make determinations on what the goals are. Can't land this until a potential 1.12.1 goes out also.

It seems 1.13 would the next release, merging.
